### ISSUE TITLE
IPv6: Web server doesn't start when no IPv4 is available

### ISSUE THREAD

**Issue by tiagogaspar8** [xbmc/xbmc#20793]
### Describe the bug
Here is a clear and concise description of what the problem is:
When Kodi is started while in a host with IPv6-Only connectiviy the web server doesn't start showing a message stating "Web Server failed to start".

When Kodi is in a system with IPv4 + IPv6 enabled:
- The server starts
- It is accessible via IPv4
- It is accessible via IPv6

When Kodi is in a system with only IPv6 enabled:
- The server doesn't start
- It is not accessible via IPv4
- It is not accessible via IPv6

## Expected Behavior
Here is a clear and concise description of what was expected to happen:
The server should be available whichever the case might be (IPv4 only, IPv4 +IPv6, or IPv6 only).

## Actual Behavior
The server is available only where IPv4 is enabled.

## Possible Fix
This might be because, at the start of the web serve, kodi checks for the existance of an IP to bind itself to and if it finds a IPv4 it starts the server at 0.0.0.0:80 and [::]:80, yet, if no IPv4 is available it exits and fails to start the server.
Other hypothese might be because kodi tries to bind to both IPv4 and IPv6, yet since IPv4 is not avilable (because it is disabled) it fails, even though IPv6 was available to bind.

### To Reproduce
1. Disable IPv4 on your device
2. Start Kodi
3. Try to reach the web server via IPv6 (you can't)

### Debuglog
The debuglog can be found here:
[kodi - Copy (2).log](https://github.com/xbmc/xbmc/files/7808114/kodi.-.Copy.2.log)

### Screenshots 
Here are some links or screenshots to help explain the problem:
![Screenshot (6)](https://user-images.githubusercontent.com/39808643/148070898-5d491375-c0c9-429d-b2c7-a813b1075a17.png)

## Additional context or screenshots (if appropriate)
Here is some additional context or explanation that might help:

The network kodi is connected to is a IPv6-Only network with IPv4 disabled at system level.

### Your Environment
Used Operating system:
 - [ ] Android
 - [ ] iOS
 - [ ] tvOS
 - [ ] Linux
 - [ ] OSX
 - [x] Windows
 - [ ] Windows UWP

 (Other OS might apply)

 - Operating system version/name: Windows 11
 - Kodi version: Latest master, the version used for the logs is the latest matrix master

*note: Once the issue is made we require you to update it with new information or Kodi versions should that be required.
Team Kodi will consider your problem report however, we will not make any promises the problem will be solved.*

**Comment by tiagogaspar8** [comment]
Hi @enen92 I have seen you help in another issue with the web server, could you give a help here? I'd do it but unfortunately I know nothing of python yet I'd like to help fix this issue!😁

**Comment by tiagogaspar8** [comment]
This is still an issue, can we please remove the stale label?

**Comment by tiagogaspar8** [comment]
This is still a valid issue.

**Comment by malard** [comment]
The web server is not the problem here — it never gets called. The failure is in the
**Windows interface enumeration**, which is hardcoded to IPv4, and in the fact that
`StartWebserver()` gates on it. That makes this a platform/network issue rather than a web
server or JSON-RPC one, so I am unassigning myself; it needs someone who owns the Windows
network layer, and it is a standalone fix rather than something that should travel on an
API PR.

Line references are against master @ 3eca6a7b3e.

### The mechanism

**1. The Windows adapter enumeration asks only for IPv4.**

```cpp
// xbmc/platform/win32/network/NetworkWin32.cpp:146
if (GetAdaptersAddresses(AF_INET, flags, nullptr, nullptr, &ulOutBufLen) != ERROR_BUFFER_OVERFLOW)
  return;
```

`flags` (`NetworkWin32.cpp:143`) does not include `GAA_FLAG_INCLUDE_ALL_INTERFACES`, and
Microsoft is explicit about what that combination means — `AF_INET` returns "only IPv4
addresses associated with **adapters with IPv4 enabled**", and under Remarks: "If the
`GAA_FLAG_INCLUDE_ALL_INTERFACES` is set, then all NDIS adapters will be retrieved even those
addresses associated with adapters **not bound to an address family specified in the *Family*
parameter**. When this flag is not set, then only the addresses that are bound to an adapter
enabled for the address family specified in the *Family* parameter are returned."

With IPv4 unbound from every adapter there is nothing to return, so the call fails with
`ERROR_NO_DATA` and `queryInterfaceList()` returns at `:147` — note this is the *first* call, so
even the `GetAdaptersAddresses() failed ...` debug line at `:162` is never reached. `m_interfaces`
is left empty, and therefore:

```cpp
// xbmc/network/Network.cpp:199
bool CNetworkBase::IsAvailable(void)
{
  const std::vector<CNetworkInterface*>& ifaces = GetInterfaceList();
  return (!ifaces.empty());
}
```

is **false** on an IPv6-only host.

**2. The web server is gated on that, and returns silently.**

```cpp
// xbmc/network/NetworkServices.cpp:669
if (!CServiceBroker::GetNetwork().IsAvailable())
  return false;
```

`CNetworkServices::Start()` turns that `false` into the 33101/33100 warning at
`NetworkServices.cpp:572-576`, and the settings toggle raises the same strings as a modal at
`:233-236` — which is the dialog in the screenshot.

### Why neither of the hypotheses in the report applies

`CWebServer::Start` already does the right thing and is simply never reached:

```cpp
// xbmc/network/WebServer.cpp:1253
int v6testSock;
if ((v6testSock = socket(AF_INET6, SOCK_STREAM, 0)) >= 0)
{
  closesocket(v6testSock);
  m_daemon_ip6 = StartMHD(MHD_USE_IPv6, port);
}
m_daemon_ip4 = StartMHD(0, port);

m_running = (m_daemon_ip6 != nullptr) || (m_daemon_ip4 != nullptr);
```

It starts the two daemons independently and reports success if **either** came up, so it does not
require IPv4 and does not fail as a pair. Once the enumeration stops hiding the interfaces, this
function should serve an IPv6-only host as-is.

### Evidence

**The attached log.** `kodi - Copy (2).log` shows `CNetworkBase::NetworkMessage - Starting network
services` at 13:35:27.596, then the EventServer thread and `JSONRPC Server: Successfully
initialized` at 13:35:27.601 — both of which start *before* the web server block
(`NetworkServices.cpp:541-555`) and neither of which has an `IsAvailable()` gate. There is then
**not one line from the `CWebserver[<port>]` logger**, and no `Cannot start Web Server on port`.
Since `CWebServer::Start` logs either "Started" or "Failed to start" unconditionally
(`WebServer.cpp:1265`/`:1268`), it was never called.

Working back through `StartWebserver()`, only two of its guards return without logging: the
`services.webserver` setting being off, and `IsAvailable()`. The setting cannot be it — the
startup toast is only queued inside `if (m_settings->GetBool(SETTING_SERVICES_WEBSERVER))`
(`:555`), and the settings-toggle path is likewise gated on that setting already reading true
(`:205`) before it calls `StartWebserver()` at `:233`. That leaves `IsAvailable()`.

**The API behaviour.** I checked directly (P/Invoke, Windows 11) that the `Family` filter keys off
whether the family is *enabled* on the adapter, not whether it currently holds an address: an
adapter here with IPv6 enabled but no IPv6 address is still enumerated under `AF_INET6`, with
`FirstUnicastAddress == NULL`. So an adapter with IPv4 *unbound* is genuinely absent from an
`AF_INET` enumeration rather than returned empty — which is what makes the list empty rather than
merely address-less.

### Platform spread

| Platform | Enumeration | Affected |
| --- | --- | --- |
| Windows (win32) | `GetAdaptersAddresses(AF_INET)` — `NetworkWin32.cpp:146`, `:152` | yes |
| Windows UWP (win10) | `GetAdaptersAddresses(AF_INET)` — `NetworkWin10.cpp:256`, `:262` | yes |
| FreeBSD | `getifaddrs`, skips `sa_family != AF_INET` — `NetworkFreebsd.cpp:194` | yes |
| Linux | parses `/proc/net/dev`, family-agnostic — `NetworkLinux.cpp:187` | no |
| Android / Darwin | not checked | ? |

So the reported symptom should be specific to Windows, UWP and FreeBSD, which fits this being
filed against Windows 11.

### This is wider than the web server

Everything else behind `IsAvailable()` is dead on such a host too: AirPlay and AirTunes
(`NetworkServices.cpp:763`, `:827`), weather (`WeatherJob.cpp:45`), RSS (`RssReader.cpp:137`),
music scraping (`MusicDatabase.cpp:4799`) and `Util.cpp:2091`. The worst of them is

```cpp
// xbmc/windows/GUIMediaWindow.cpp:1850
while (!CServiceBroker::GetNetwork().IsAvailable())
```

— browsing any network source opens the "Loading Directory" progress dialog and spins there
forever, dismissable only by cancelling. Separately, `GetCurrentIPAddress()` returns `""` for every
interface, so the advertised UPnP address (`UPnP.cpp:536`, `UPnPRenderer.cpp:455`), the System Info
IP (`SystemInfo.cpp:331`) and `xbmc.getIPAddress()` (`ModuleXbmc.cpp:267`) are all blank.

### Suggested fix

1. Switch `queryInterfaceList()` to `AF_UNSPEC` — `NetworkWin32.cpp:146`/`:152` and
   `NetworkWin10.cpp:256`/`:262`.
2. **Not on its own, though.** `CNetworkInterfaceWin32::GetCurrentIPAddress()`,
   `GetCurrentNetmask()` and `GetCurrentDefaultGateway()` (`NetworkWin32.cpp:74-99`) read only the
   *head* of `FirstUnicastAddress` / `FirstGatewayAddress`, so `AF_UNSPEC` alone would start
   handing IPv6 literals to dual-stack users depending on how Windows orders the list. They should
   walk the list preferring `AF_INET` and fall back to the first `AF_INET6`. That keeps dual-stack
   behaviour byte-identical and confines the change to hosts that have no IPv4 at all.
   (`GetCurrentNetmask` already has an `AF_INET6` branch at `:87-90` returning a bare prefix
   length, so this was half-anticipated.)

That alone fixes the report: `IsAvailable()` becomes true, `StartWebserver()` proceeds, and
`CWebServer::Start` brings up the IPv6 daemon.

Worth considering separately: **drop the `IsAvailable()` guard at `NetworkServices.cpp:669`
entirely.** A listening server does not need an enumerable interface — MHD binds the wildcard
address — and neither the EventServer nor the JSON-RPC server has such a gate, which is exactly
why both started in the reporter's log while the web server did not. That would make the web
server robust against any enumeration failure, not just this one.

Two things a fix should be careful of. `CNetworkBase::HasInterfaceForIP`
(`Network.cpp:177-197`) runs `inet_addr()` over `GetCurrentIPAddress()` and `GetCurrentNetmask()`;
once those can be IPv6 text and a bare prefix length, `inet_addr("64")` yields `0.0.0.64` rather
than an error, so it should return false outright when the interface has no IPv4 address. And UPnP
presumably wants an IPv4 literal specifically rather than whatever the interface reports first —
that may argue for an explicit family preference on the accessor instead of an implicit
IPv4-first rule.

### Caveat

I have **not** reproduced this: it needs a host with IPv4 unbound from every adapter, which I do
not have. The mechanism above is a code-reading diagnosis corroborated by the attached log and by
the `GetAdaptersAddresses` behaviour test, and it accounts for the report — including the total
absence of web server logging — without further assumptions, but a live before/after on an
IPv6-only host is still what a fix should be verified against. @tiagogaspar8, if you still have
that setup, the check that would settle it is whether the System Info IP address field is blank
and whether browsing a network source hangs on "Loading Directory"; both should follow from the
same empty interface list.

**Comment by tiagogaspar8** [comment]
Hi @malard,

Great analysis! I actually didn't try this on Linux, I assumed it was a general problem.
If I need to update the issue name/description let me know.

Thanks!
